### PR TITLE
Nosp date saving

### PR DESCRIPTION
**Pull request by elena-vi**
https://hackney.atlassian.net/secure/RapidBoard.jspa?rapidView=22&projectKey=MA&modal=detail&selectedIssue=MA-134

Officers need to see when a nosp has been served, also we need these dates for prioritisation

### CODE REVIEW

File: lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb [hunk 1]
```diff
@@ -88,6 +97,18 @@ def balance
           attributes.fetch(:current_balance).to_f
         end
 
+        def nosp_served_date
+          return nil if attributes.fetch(:nosp_served_date).nil?
```
Comment (kjdchapman) [inline lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb]: I think `fetch` will throw an exception if it doesn't find the key.

File: spec/lib/hackney/income/tenancy_prioritiser/universal_housing_criteria_spec.rb [hunk 2]
```diff
@@ -6,6 +6,11 @@
   let(:universal_housing_client) { Hackney::UniversalHousing::Client.connection }
 
   let(:tenancy_ref) { '000015/01' }
+
+  let(:nosp_notice_served_date) { '2005-12-13 12:43:10' }
+
+  let(:nosp_notice_expiry_date) { '2015-12-13 12:43:10' }
```
Comment (kjdchapman) [inline spec/lib/hackney/income/tenancy_prioritiser/universal_housing_criteria_spec.rb]: I'd use different dates for these two, to let me know if I accidentally mix them up somewhere else.

File: spec/lib/hackney/income/tenancy_prioritiser/universal_housing_criteria_spec.rb [hunk 3]
```diff
@@ -22,6 +27,22 @@
...
+
+    it 'returns the nosp served date' do
+      expect(subject).to eq(nosp_notice_served_date.to_date)
+    end
+  end
+
+  describe '#nosp_expiry_date' do
+    subject { criteria.nosp_expiry_date }
+
+    it 'returns the  nosp expiry date' do
```
Comment (kjdchapman) [inline spec/lib/hackney/income/tenancy_prioritiser/universal_housing_criteria_spec.rb]: nit: extra space
```suggestion
    it 'returns the nosp expiry date' do
```

File: lib/hackney/income/stored_tenancies_gateway.rb [hunk 4]
```diff
@@ -34,6 +34,9 @@ def store_tenancy(tenancy_ref:, priority_band:, priority_score:, criteria:, weig
               active_agreement: criteria.active_agreement?,
               broken_court_order: criteria.broken_court_order?,
               nosp_served: criteria.nosp_served?,
+              nosp_served_date: criteria.nosp_served_date,
```
Comment (kjdchapman) [inline lib/hackney/income/stored_tenancies_gateway.rb]: Is there any duplication between `nosp_served` and `nosp_served_date`? Can we infer the former from the latter?

Probably not something to address in this PR either way.

Comment (elena-vi) [reply lib/hackney/income/stored_tenancies_gateway.rb]: `nosp_served` returns a boolean, also that's used in the current prioritisation so I wanted to keep that in for now

File: spec/lib/hackney/income/tenancy_prioritiser/universal_housing_criteria_spec.rb (hunk 2 above)
Comment (elena-vi) [reply spec/lib/hackney/income/tenancy_prioritiser/universal_housing_criteria_spec.rb]:
```suggestion
  let(:nosp_notice_expiry_date) { '2019-10-20 14:31:12' }
```

File: lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb [hunk 5]
```diff
@@ -88,6 +97,18 @@ def balance
           attributes.fetch(:current_balance).to_f
         end
 
+        def nosp_served_date
+          return nil if attributes.dig(:nosp_served_date).nil?
```
Comment (kjdchapman) [inline lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb]: `dig` with one argument can just be hash access, if you want. Also, I think `nil` is returned implicitly if you don't specify a return value?

```suggestion
          return if attributes[:nosp_served_date].nil?
```

File: lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb [hunk 6]
```diff
@@ -88,6 +97,18 @@ def balance
           attributes.fetch(:current_balance).to_f
         end
 
+        def nosp_served_date
+          return if attributes[:nosp_served_date].nil?
+
+          attributes.fetch(:nosp_served_date).to_date
```
Comment (elena-vi) [inline lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb]:
```suggestion
          attributes[:nosp_served_date].to_date
```

File: lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb [hunk 7]
```diff
@@ -88,6 +97,18 @@ def balance
...
+        end
+
+        def nosp_expiry_date
+          return nil if attributes.dig(:nosp_served_date).nil?
```
Comment (elena-vi) [inline lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb]:
```suggestion
          return nil if attributes[:nosp_served_date].nil?
```

File: lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb [hunk 8]
```diff
@@ -88,6 +97,18 @@ def balance
...
+
+          attributes.fetch(:nosp_expiry_date).to_date
```
Comment (elena-vi) [inline lib/hackney/income/tenancy_prioritiser/universal_housing_criteria.rb]:
```suggestion
          attributes[:nosp_expiry_date].to_date
```

**Pull request merged**